Improve syntax error for misplaced lifetime quantifier (`for<'a>`) outside of APIT type

### Code

```Rust
fn foo(_f: for<'a> impl FnOnce(&'a str)) {}
```


### Current output

```Shell
error: expected identifier, found keyword `impl`
 --> src/lib.rs:1:20
  |
1 | fn foo(_f: for<'a> impl FnOnce(&'a str)) {}
  |                    ^^^^ expected identifier, found keyword

error: unexpected lifetime `'a` in pattern
 --> src/lib.rs:1:33
  |
1 | fn foo(_f: for<'a> impl FnOnce(&'a str)) {}
  |                                 ^^ help: remove the lifetime

error: expected one of `:` or `|`, found `)`
 --> src/lib.rs:1:40
  |
1 | fn foo(_f: for<'a> impl FnOnce(&'a str)) {}
  |                                        ^ expected one of `:` or `|`

error: expected one of `(`, `)`, `+`, `,`, `::`, or `<`, found `FnOnce`
 --> src/lib.rs:1:25
  |
1 | fn foo(_f: for<'a> impl FnOnce(&'a str)) {}
  |                        -^^^^^^ expected one of `(`, `)`, `+`, `,`, `::`, or `<`
  |                        |
  |                        help: missing `,`

error[E0405]: cannot find trait `r#impl` in this scope
 --> src/lib.rs:1:20
  |
1 | fn foo(_f: for<'a> impl FnOnce(&'a str)) {}
  |                    ^^^^ not found in this scope

error[E0782]: trait objects must include the `dyn` keyword
 --> src/lib.rs:1:12
  |
1 | fn foo(_f: for<'a> impl FnOnce(&'a str)) {}
  |            ^^^^^^^^^^^^
  |
help: add `dyn` keyword before this trait
  |
1 | fn foo(_f: dyn for<'a> impl FnOnce(&'a str)) {}
  |            +++
```


### Desired output

```Shell
error: a lifetime quantifier (`for<'a>`) may not be used here
 --> src/lib.rs:1:20
  |
1 | fn foo(_f: for<'a> impl FnOnce(&'a str)) {}
  |            ^^^^^^^ this quantifier
  |
note: lifetime quantifiers may only be used before `fn` pointer types and within `impl Trait` types
help: place this `for<'a>` inside the `impl`:
  |
1 | fn foo(_f: impl for<'a> FnOnce(&'a str)) {}
  |                 +++++++
```


### Rationale and extra context

I don't remember what the official name for `for<'a>` syntax in general is, so I've used "quantifier" in the proposed message.

The main thing I'm filing the issue for is that there's a large cascade of syntax errors, none of which explain the actual problem (the leading "expected identifier" is particularly bad), and which seem to be due to poor recovery (starting to parse the rest of the type as a new function parameter, as if there was a comma); it would be nice to have more concise and more helpful output, particularly because `for<'a>` is a superficially simple "modifier" syntax that _does_ appear in multiple places in types, just not this one.

I hope that it will be simple to fix by making `for<'a>` something that is always parsed but not always permitted.

Also, rust-analyzer's current diagnostic for this is significantly better just by being a basic parse error: "expected a function pointer or path". It doesn't point the user to the right fix, but it does actually mention what the non-syntax-error cases would be: a `fn` pointer with lifetimes — or a pre-2021-unmarked-`dyn` (sigh).

### Other cases

_No response_

### Anything else?

Occurs on stable 1.73 and  1.76.0-nightly (2023-11-12 2b603f95a48f10f931a6)

--- a/compiler/rustc_parse/messages.ftl
+++ b/compiler/rustc_parse/messages.ftl
@@ -739,6 +739,9 @@ parse_trailing_vert_not_allowed = a trailing `|` is not allowed in an or-pattern
 parse_trait_alias_cannot_be_auto = trait aliases cannot be `auto`
 parse_trait_alias_cannot_be_unsafe = trait aliases cannot be `unsafe`
 
+parse_transpose_dyn_or_impl = `for<...>` expected after `{$kw}`, not before
+    .suggestion = move `{$kw}` before the `for<...>`
+
 parse_type_ascription_removed =
     if you meant to annotate an expression with a type, the type ascription syntax has been removed, see issue #101728 <https://github.com/rust-lang/rust/issues/101728>
 

--- a/compiler/rustc_parse/src/errors.rs
+++ b/compiler/rustc_parse/src/errors.rs
@@ -2828,3 +2828,23 @@ pub(crate) struct GenericArgsInPatRequireTurbofishSyntax {
     )]
     pub suggest_turbofish: Span,
 }
+
+#[derive(Diagnostic)]
+#[diag(parse_transpose_dyn_or_impl)]
+pub(crate) struct TransposeDynOrImpl<'a> {
+    #[primary_span]
+    pub span: Span,
+    pub kw: &'a str,
+    #[subdiagnostic]
+    pub sugg: TransposeDynOrImplSugg<'a>,
+}
+
+#[derive(Subdiagnostic)]
+#[multipart_suggestion(parse_suggestion, applicability = "machine-applicable")]
+pub(crate) struct TransposeDynOrImplSugg<'a> {
+    #[suggestion_part(code = "")]
+    pub removal_span: Span,
+    #[suggestion_part(code = "{kw} ")]
+    pub insertion_span: Span,
+    pub kw: &'a str,
+}

--- a/compiler/rustc_parse/src/parser/ty.rs
+++ b/compiler/rustc_parse/src/parser/ty.rs
@@ -287,6 +287,7 @@ impl<'a> Parser<'a> {
             // Function pointer type
             self.parse_ty_bare_fn(lo, ThinVec::new(), None, recover_return_sign)?
         } else if self.check_keyword(kw::For) {
+            let for_span = self.token.span;
             // Function pointer type or bound list (trait object type) starting with a poly-trait.
             //   `for<'lt> [unsafe] [extern "ABI"] fn (&'lt S) -> T`
             //   `for<'lt> Trait1<'lt> + Trait2 + 'a`
@@ -299,9 +300,42 @@ impl<'a> Parser<'a> {
                     recover_return_sign,
                 )?
             } else {
-                let path = self.parse_path(PathStyle::Type)?;
-                let parse_plus = allow_plus == AllowPlus::Yes && self.check_plus();
-                self.parse_remaining_bounds_path(lifetime_defs, path, lo, parse_plus)?
+                // Try to recover `for<'a> dyn Trait` or `for<'a> impl Trait`.
+                if self.may_recover()
+                    && (self.eat_keyword_noexpect(kw::Impl) || self.eat_keyword_noexpect(kw::Dyn))
+                {
+                    let kw = self.prev_token.ident().unwrap().0.name;
+                    let mut err = self.sess.create_err(errors::TransposeDynOrImpl {
+                        span: self.prev_token.span,
+                        kw: kw.as_str(),
+                        sugg: errors::TransposeDynOrImplSugg {
+                            removal_span: self.prev_token.span.with_hi(self.token.span.lo()),
+                            insertion_span: for_span.shrink_to_lo(),
+                            kw: kw.as_str(),
+                        },
+                    });
+                    let path = self.parse_path(PathStyle::Type)?;
+                    let parse_plus = allow_plus == AllowPlus::Yes && self.check_plus();
+                    let kind =
+                        self.parse_remaining_bounds_path(lifetime_defs, path, lo, parse_plus)?;
+                    // Take the parsed bare trait object and turn it either
+                    // into a `dyn` object or an `impl Trait`.
+                    let kind = match (kind, kw) {
+                        (TyKind::TraitObject(bounds, _), kw::Dyn) => {
+                            TyKind::TraitObject(bounds, TraitObjectSyntax::Dyn)
+                        }
+                        (TyKind::TraitObject(bounds, _), kw::Impl) => {
+                            TyKind::ImplTrait(ast::DUMMY_NODE_ID, bounds)
+                        }
+                        _ => return Err(err),
+                    };
+                    err.emit();
+                    kind
+                } else {
+                    let path = self.parse_path(PathStyle::Type)?;
+                    let parse_plus = allow_plus == AllowPlus::Yes && self.check_plus();
+                    self.parse_remaining_bounds_path(lifetime_defs, path, lo, parse_plus)?
+                }
             }
         } else if self.eat_keyword(kw::Impl) {
             self.parse_impl_ty(&mut impl_dyn_multi)?

--- a/tests/ui/parser/recover-hrtb-before-dyn-impl-kw.rs
+++ b/tests/ui/parser/recover-hrtb-before-dyn-impl-kw.rs
@@ -1,0 +1,9 @@
+trait Trait {}
+
+fn test(_: &for<'a> dyn Trait) {}
+//~^ ERROR `for<...>` expected after `dyn`, not before
+
+fn test2(_: for<'a> impl Trait) {}
+//~^ ERROR `for<...>` expected after `impl`, not before
+
+fn main() {}

--- a/tests/ui/parser/recover-hrtb-before-dyn-impl-kw.stderr
+++ b/tests/ui/parser/recover-hrtb-before-dyn-impl-kw.stderr
@@ -1,0 +1,26 @@
+error: `for<...>` expected after `dyn`, not before
+  --> $DIR/recover-hrtb-before-dyn-impl-kw.rs:3:21
+   |
+LL | fn test(_: &for<'a> dyn Trait) {}
+   |                     ^^^
+   |
+help: move `dyn` before the `for<...>`
+   |
+LL - fn test(_: &for<'a> dyn Trait) {}
+LL + fn test(_: &dyn for<'a> Trait) {}
+   |
+
+error: `for<...>` expected after `impl`, not before
+  --> $DIR/recover-hrtb-before-dyn-impl-kw.rs:6:21
+   |
+LL | fn test2(_: for<'a> impl Trait) {}
+   |                     ^^^^
+   |
+help: move `impl` before the `for<...>`
+   |
+LL - fn test2(_: for<'a> impl Trait) {}
+LL + fn test2(_: impl for<'a> Trait) {}
+   |
+
+error: aborting due to 2 previous errors
+

